sbin/fsck/fsck.8
.\"	$NetBSD: fsck.8,v 1.19 1999/03/10 00:08:33 erh Exp $
.\"
.\" Copyright (c) 1996 Christos Zoulas.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by Christos Zoulas.
.\" 4. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
.\" INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
.\" NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
.\" THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd March 5, 2019
.Dt FSCK 8
.Os
.Sh NAME
.Nm fsck
.Nd file system consistency check and interactive repair
.Sh SYNOPSIS
.Nm
.Op Fl Cdfnpvy
.Op Fl B | F
.Op Fl T Ar fstype : Ns Ar fsoptions
.Op Fl t Ar fstype
.Op Fl c Ar fstab
.Oo Ar special | node Oc ...
.Sh DESCRIPTION
The
.Nm
utility invokes file system-specific programs to check
the special devices listed in the
.Xr fstab 5
file or in the command line for consistency.
.Pp
It is normally used in the script
.Pa /etc/rc
during automatic reboot.
Traditionally,
.Nm
is invoked before the file systems are mounted
and all checks are done to completion at that time.
If background checking is available,
.Nm
is invoked twice.
It is first invoked at the traditional time,
before the file systems are mounted, with the
.Fl F
flag to do checking on all the file systems
that cannot do background checking.
It is then invoked a second time,
after the system has completed going multiuser, with the
.Fl B
flag to do checking on all the file systems
that can do background checking.
Unlike the foreground checking,
the background checking is started asynchronously
so that other system activity can proceed
even on the file systems that are being checked.
.Pp
If no file systems are specified,
.Nm
reads the table
.Pa /etc/fstab
to determine which file systems to check.
Only partitions in
.Pa /etc/fstab
that are mounted
.Dq rw ,
.Dq rq
or
.Dq ro
and that have non-zero pass number are checked.
File systems with pass number 1 (normally just the root file system)
are always checked one at a time.
.Pp
If not in preen mode, the remaining entries are checked in order of
increasing pass number one at a time.
This is needed when interaction with
.Nm
is required.
.Pp
In preen mode, after pass 1 completes, all remaining file systems are checked,
in pass number order running one process per disk drive in parallel for each
pass number in increasing order.
.Pp
In other words: In preen mode all pass 1 partitions are checked sequentially.
Next all pass 2 partitions are checked in parallel, one process per disk drive.
Next all pass 3 partitions are checked in parallel, one process per disk drive.
etc.
.Pp
The disk drive containing each file system is inferred from the shortest prefix
of the device name that ends in a digit; the remaining characters are assumed
to be the partition and slice designators.
.Pp
If the
.Fl t
or
.Fl T
flags are not specified,
.Nm
will attempt to determine the file system type and call the
appropriate file system check utility.
Failure to detect the file system type will cause
.Nm
to fail with a message that the partition has an unknown file system type.
.Pp
The options are as follows:
.Bl -tag -width indent
.It Fl C
Check if the
.Dq clean
flag is set in the superblock and skip file system checks if file system was
properly dismounted and marked clean.
.It Fl c Ar fstab
Specify the
.Pa fstab
file to use.
.It Fl d
Debugging mode.
Just print the commands without executing them.
Available
only if
.Nm
is compiled to support it.
.It Fl f
Force checking of file systems.
Running
.Dq Li fsck -f
ignores the journal and does a full
consistency check of the disk so will find and fix the errors about which the
journal is unaware.
.It Fl n
Causes
.Nm
to assume no as the answer to all operator questions, except "CONTINUE?".
.It Fl p
Enter preen mode.
In preen mode, only a restricted class of innocuous
file system inconsistencies will be corrected.
If unexpected inconsistencies caused by hardware or
software failures are encountered, the check program
will exit with a failure.
See the manual pages for the individual check programs
for a list of the sorts of failures that they correct
when running in preen mode.
.It Fl F
Run in foreground mode.
The check program for each file system is invoked with the
.Fl F
flag to determine whether it wishes to run as part of
the boot up sequence,
or if it is able to do its job in background after the
system is up and running.
A non-zero exit code indicates that it wants to run in foreground
and the check program is invoked.
A zero exit code indicates that it is able to run later in background
and just a deferred message is printed.
.It Fl B
Run in background mode.
The check program for each file system is invoked with the
.Fl F
flag to determine whether it wishes to run as part of
the boot up sequence,
or if it is able to do its job in background after the
system is up and running.
A non-zero exit code indicates that it wanted to run in foreground
which is assumed to have been done, so the file system is skipped.
A zero exit code indicates that it is able to run in background
so the check program is invoked with the
.Fl B
flag to indicate that a check on the active file system should be done.
When running in background mode,
only one file system at a time will be checked.
Note that background
.Nm
is limited to checking for only the most commonly occurring
file system abnormalities.
Under certain circumstances,
some errors can escape background
.Nm .
It is recommended that you perform foreground
.Nm
on your systems periodically and whenever you encounter
file-system\-related panics.
.It Fl t Ar fstype
Invoke
.Nm
only for the comma separated list of file system types.
If the
list starts with
.Dq no
then invoke
.Nm
for the file system types that are not specified in the list.
.It Fl v
Print the commands before executing them.
.It Fl y
Causes
.Nm
to assume yes
as the answer to all operator questions.
.It Fl T Ar fstype : Ns Ar fsoptions
List of comma separated file system specific options for the specified
file system type, in the same format as
.Xr mount 8 .
.El
.Sh FILES
.Bl -tag -width /etc/fstab -compact
.It Pa /etc/fstab
file system table
.El
.Sh SEE ALSO
.Xr fstab 5 ,
.Xr fsck_ffs 8 ,
.Xr fsck_msdosfs 8 ,
.Xr mount 8
.Sh HISTORY
A
.Nm
utility appeared in
.Bx 4.0 .
It was reimplemented as a filesystem independent wrapper in
.Nx 1.3
and first appeared in
.Fx 5.0 .
The original filesystem specific utility became
.Xr fsck_ffs 8
at this point.


NEWS
Dec 07, 2022: libarchive 3.6.2 released

Apr 08, 2022: libarchive 3.6.1 released

Feb 09, 2022: libarchive 3.6.0 released

Feb 08, 2022: libarchive 3.5.3 released

Aug 22, 2021: libarchive 3.5.2 released

Dec 26, 2020: libarchive 3.5.1 released

Dec 01, 2020: libarchive 3.5.0 released

Oct 14, 2020: Support for system extended attributes

May 20, 2020: libarchive 3.4.3 released

Apr 30, 2020: Support for pzstd compressed files

Apr 16, 2020: Support for RHT.security.selinux tar extended attribute

Feb 11, 2020: libarchive 3.4.2 released

Jan 23, 2020: Important fixes for writing XAR archives

Jan 20, 2020: New tar option: --safe-writes (atomical file extraction)

Jan 03, 2020: Support mbed TLS (PolarSSL) as optional crypto provider

Dec 30, 2019: libarchive 3.4.1 released

Dec 11, 2019: New pax write option "xattrhdr"

Nov 17, 2019: Unicode filename support for reading lha/lzh archives

Jun 11, 2019: libarchive 3.4.0 released

May 18, 2019: Fixes for reading Android APK and JAR archives

Apr 16, 2019: Support for non-recursive list and extract

Apr 14, 2019: New tar option: --exclude-vcs

Mar 27, 2019: Support for file and directory symlinks on Windows

Mar 12, 2019: Important fixes for storing file attributes and flags

Jan 20, 2019: Support for xz, lzma, ppmd8 and bzip2 decompression in ZIP files

Oct 06, 2018: RAR 5.0 reader

Sep 03, 2018: libarchive 3.3.3 released

Jul 19, 2018: Avoid super-linear slowdown on malformed mtree files

Jan 27, 2018: Many fixes for building with Visual Studio

Oct 19, 2017: NO_OVERWRITE doesn't change existing directory attributes

Aug 12, 2017: New support for Zstandard read and write filters

Jul 09, 2017: libarchive 3.3.2 released

Mar 16, 2017: NFSv4 ACL support for Linux (librichacl)

Feb 26, 2017: libarchive 3.3.1 released
    Security & Feature release

Feb 19, 2017: libarchive 3.3.0 released
    Security & Feature release

Jan 29, 2017: Limited NFSv4 ACL support for Mac OS (Darwin)

Jan 10, 2017: POSIX.1e and NFSv4 ACL support for Solaris and derivates

Dec 27, 2016: NFSv4 ACL read and write support for pax
    Deprecated functions: archive_entry_acl_text(), archive_entry_acl_text_w()

Nov, 2016: libarchive is now being tested by the OSS-Fuzz project

Oct 26, 2016: Remove liblzmadec support

Oct 23, 2016: libarchive 3.2.2 released
    Security release

Jun 20, 2016: libarchive 3.2.1 released
    This fixes a handful of security and other critical issues with 3.2.0

May 01, 2016: libarchive 3.2.0 released

Apr 09, 2016: libarchive 3.1.901a released
    Another test release in preparation for 3.2.0

Feb 13, 2016: libarchive 3.1.900a released
    This is a test release in preparation for 3.2.0

Oct 21, 2015: Preliminary port to OSF

Apr 11, 2015: libarchive's issue tracker is now hosted at GitHub.
    https://github.com/libarchive/libarchive/issues

Early 2015:  Many fixes to crash and overflow bugs thanks to Hanno Boeck

Oct 13, 2014: Zip encryption and decryption support

Aug 13, 2014: Add support for lz4 compression.

Jun 10, 2014: Add warc format support

May  3, 2014: Add experimental Zip streaming extension

Apr  6, 2014: Add bsdcat command-line tool

Jan 12, 2014: Add Zip64 support

Dec  1, 2013: Rewrite Zip write logic

Jul  1, 2013: Add ability to detect encrypted entries for many formats
     (This does not add the ability to *decrypt* those entries, however)

Feb 23, 2013: "raw" write support added

Feb 09, 2013: libarchive 3.1.2 released

Jan 28, 2013: libarchive's new website moved to http://www.libarchive.org.

Jan 13, 2013: libarchive 3.1.1 released

Jan 13, 2013: libarchive 3.1.0 released

Dec 07, 2012: Implement functions to manually set the format and filters used.
    
Nov 11, 2012: Add support for __MACOSX directory in Zip archives, which resource
    forks are stored in.

Oct 20, 2012: Add support for writing v7 tar format.

Oct 09, 2012: Add support for grzip compression.

Oct 07, 2012: Introduce b64encode filter.    
Oct 07, 2012: Introduce uuencode filter.

Oct 06, 2012: Add support for lzop.
    
Sep 27, 2012: Implement function used to seek within data blocks.
    (Currently only supported for uncompressed RAR archives).
    
Apr 22, 2012: Add basic archive read and write filter support for lrzip.

Mar 27, 2012: libarchive 3.0.4 released

Feb 05, 2012: libarchive development now hosted at GitHub.
    http://libarchive.github.com/
Feb 05, 2012: libarchive's issue tracker remains at Google Code.
    http://code.google.com/p/libarchive/issues/list
Feb 05, 2012: libarchive's mailing lists remain at Google Groups.

Dec 24, 2011: libarchive 3.0.2 released
Dec 23, 2011: Various fixes merged from FreeBSD
Dec 23, 2011: Symlink support in Zip reader and writer
Dec 23, 2011: Robustness fixes to 7Zip reader

Nov 27, 2011: libarchive 3.0.1b released

Nov 26, 2011: 7Zip reader
Nov 26, 2011: Small fixes to ISO and Zip to improve robustness with corrupted input
Nov 24, 2011: Improve streaming Zip reader's support for uncompressed entries
Nov 20, 2011: New seeking Zip reader supports SFX Zip archives
Nov 20, 2011: Build fixes on Windows

Nov 13, 2011: libarchive 3.0.0a released

Nov 06, 2011: Update shared-library version calculations for libarchive 3.x
Sep 04, 2011: Fix tar -s; follow GNU tar for controlling hardlink/symlink substitutions
Aug 18, 2011: Fix reading ISO images built by NetBSD's mkisofs
Aug 15, 2011: Old archive_read_support_compression_XXX functions are deprecated and
   will disappear in libarchive 4.0.
Jun 26, 2011: RAR reader
Jun 16, 2011: Add tar:compat-2x option to emulate broken libarchive 2.x
   handling of pax UTF-8 headers
Apr 25, 2011: Refactor read_open() into a collection of single-item setters;
   support the old interfaces as wrappers
Apr 12, 2011: Split disk writer into separate POSIX and Windows implementations
Apr 10, 2011: Improvements to character translations on Windows.
Mar 30, 2011: More work to return errors instead of calling abort()

sbin/fsck/fsck.c
/*	$NetBSD: fsck.c,v 1.30 2003/08/07 10:04:15 agc Exp $	*/

/*-
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Copyright (c) 1996 Christos Zoulas. All rights reserved.
 * Copyright (c) 1980, 1989, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * From: @(#)mount.c	8.19 (Berkeley) 4/19/94
 * From: $NetBSD: mount.c,v 1.24 1995/11/18 03:34:29 cgd Exp 
 * $NetBSD: fsck.c,v 1.30 2003/08/07 10:04:15 agc Exp $
 */

#include <sys/cdefs.h>
#include <sys/param.h>
#include <sys/mount.h>
#include <sys/queue.h>
#include <sys/wait.h>
#include <sys/disk.h>
#include <sys/ioctl.h>

#include <ctype.h>
#include <err.h>
#include <fstab.h>
#include <fcntl.h>
#include <mntopts.h>
#include <paths.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "fsutil.h"

static enum { IN_LIST, NOT_IN_LIST } which = NOT_IN_LIST;

static TAILQ_HEAD(fstypelist, entry) opthead, selhead;

struct entry {
	char *type;
	char *options;
	TAILQ_ENTRY(entry) entries;
};

static char *options = NULL;
static int flags = 0;
static int forceflag = 0;

static int checkfs(const char *, const char *, const char *, const char *, pid_t *);
static int selected(const char *);
static void addoption(char *);
static const char *getoptions(const char *);
static void addentry(struct fstypelist *, const char *, const char *);
static void maketypelist(char *);
static void catopt(char **, const char *);
static void mangle(char *, int *, const char ** volatile *, int *);
static const char *getfstype(const char *);
static void usage(void) __dead2;
static int isok(struct fstab *);

static struct {
	const char *ptype;
	const char *name;
} ptype_map[] = {
	{ "ufs",	"ffs" },
	{ "ffs",	"ffs" },
	{ "fat",	"msdosfs" },
	{ "efi",	"msdosfs" },
	{ NULL,		NULL },
};

int
main(int argc, char *argv[])
{
	struct fstab *fs;
	int i, rval = 0;
	const char *vfstype = NULL;
	char globopt[3];
	const char *etc_fstab;

	globopt[0] = '-';
	globopt[2] = '\0';

	TAILQ_INIT(&selhead);
	TAILQ_INIT(&opthead);

	etc_fstab = NULL;
	while ((i = getopt(argc, argv, "BCdvpfFnyl:t:T:c:")) != -1)
		switch (i) {
		case 'B':
			if (flags & CHECK_BACKGRD)
				errx(1, "Cannot specify -B and -F.");
			flags |= DO_BACKGRD;
			break;

		case 'd':
			flags |= CHECK_DEBUG;
			break;

		case 'v':
			flags |= CHECK_VERBOSE;
			break;

		case 'F':
			if (flags & DO_BACKGRD)
				errx(1, "Cannot specify -B and -F.");
			flags |= CHECK_BACKGRD;
			break;

		case 'p':
			flags |= CHECK_PREEN;
			/*FALLTHROUGH*/
		case 'C':
			flags |= CHECK_CLEAN;
			/*FALLTHROUGH*/
		case 'n':
		case 'y':
			globopt[1] = i;
			catopt(&options, globopt);
			break;

		case 'f':
			forceflag = 1;
			globopt[1] = i;
			catopt(&options, globopt);
			break;

		case 'l':
			warnx("Ignoring obsolete -l option\n");
			break;

		case 'T':
			if (*optarg)
				addoption(optarg);
			break;

		case 't':
			if (!TAILQ_EMPTY(&selhead))
				errx(1, "only one -t option may be specified.");

			maketypelist(optarg);
			vfstype = optarg;
			break;

		case 'c':
			etc_fstab = optarg;
			break;

		case '?':
		default:
			usage();
			/* NOTREACHED */
		}

	argc -= optind;
	argv += optind;

	if (etc_fstab != NULL)
		setfstab(etc_fstab);

	if (argc == 0)
		return checkfstab(flags, isok, checkfs);

#define	BADTYPE(type)							\
	(strcmp(type, FSTAB_RO) &&					\
	    strcmp(type, FSTAB_RW) && strcmp(type, FSTAB_RQ))


	for (; argc--; argv++) {
		const char *spec, *mntpt, *type, *cp;
		char device[MAXPATHLEN];
		struct statfs *mntp;

		mntpt = NULL;
		spec = *argv;
		cp = strrchr(spec, '/');
		if (cp == NULL) {
			(void)snprintf(device, sizeof(device), "%s%s",
				_PATH_DEV, spec);
			spec = device;
		}
		mntp = getmntpoint(spec);
		if (mntp != NULL) {
			spec = mntp->f_mntfromname;
			mntpt = mntp->f_mntonname;
		}
		if ((fs = getfsfile(spec)) == NULL &&
		    (fs = getfsspec(spec)) == NULL) {
			if (vfstype == NULL)
				vfstype = getfstype(spec);
			if (vfstype == NULL)
				vfstype = "ufs";
			type = vfstype;
			devcheck(spec);
		} else {
			spec = fs->fs_spec;
			type = fs->fs_vfstype;
			mntpt = fs->fs_file;
			if (BADTYPE(fs->fs_type))
				errx(1, "%s has unknown file system type.",
				    spec);
		}
		if ((flags & CHECK_BACKGRD) &&
		    checkfs(type, spec, mntpt, "-F", NULL) == 0) {
			printf("%s: DEFER FOR BACKGROUND CHECKING\n", *argv);
			continue;
		}
		if ((flags & DO_BACKGRD) && forceflag == 0 &&
		    checkfs(type, spec, mntpt, "-F", NULL) != 0)
			continue;

		rval |= checkfs(type, spec, mntpt, NULL, NULL);
	}

	return rval;
}


static int
isok(struct fstab *fs)
{
	int i;

	if (fs->fs_passno == 0)
		return (0);
	if (BADTYPE(fs->fs_type))
		return (0);
	if (!selected(fs->fs_vfstype))
		return (0);
	/* If failok, always check now */
	if (getfsopt(fs, "failok"))
		return (1);
	/*
	 * If the -B flag has been given, then process the needed
	 * background checks. Background checks cannot be run on
	 * file systems that will be mounted read-only or that were
	 * not mounted at boot time (typically those marked `noauto').
	 * If these basic tests are passed, check with the file system
	 * itself to see if it is willing to do background checking
	 * by invoking its check program with the -F flag.
	 */
	if (flags & DO_BACKGRD) {
		if (!strcmp(fs->fs_type, FSTAB_RO))
			return (0);
		if (getmntpoint(fs->fs_spec) == NULL)
			return (0);
		if (checkfs(fs->fs_vfstype, fs->fs_spec, fs->fs_file, "-F", 0))
			return (0);
		return (1);
	}
	/*
	 * If the -F flag has been given, then consider deferring the
	 * check to background. Background checks cannot be run on
	 * file systems that will be mounted read-only or that will
	 * not be mounted at boot time (e.g., marked `noauto'). If
	 * these basic tests are passed, check with the file system
	 * itself to see if it is willing to defer to background
	 * checking by invoking its check program with the -F flag.
	 */
	if ((flags & CHECK_BACKGRD) == 0 || !strcmp(fs->fs_type, FSTAB_RO))
		return (1);
	for (i = strlen(fs->fs_mntops) - 6; i >= 0; i--)
		if (!strncmp(&fs->fs_mntops[i], "noauto", 6))
			break;
	if (i >= 0)
		return (1);
	if (checkfs(fs->fs_vfstype, fs->fs_spec, fs->fs_file, "-F", NULL) != 0)
		return (1);
	printf("%s: DEFER FOR BACKGROUND CHECKING\n", fs->fs_spec);
	return (0);
}


static int
checkfs(const char *pvfstype, const char *spec, const char *mntpt,
    const char *auxopt, pid_t *pidp)
{
	const char ** volatile argv;
	pid_t pid;
	int argc, i, status, maxargc;
	char *optbuf, execbase[MAXPATHLEN];
	char *vfstype = NULL;
	const char *extra = NULL;

#ifdef __GNUC__
	/* Avoid vfork clobbering */
	(void) &optbuf;
	(void) &vfstype;
#endif
	/*
	 * We convert the vfstype to lowercase and any spaces to underscores
	 * to not confuse the issue
	 *
	 * XXX This is a kludge to make automatic filesystem type guessing
	 * from the disklabel work for "4.2BSD" filesystems.  It does a
	 * very limited subset of transliteration to a normalised form of
	 * filesystem name, and we do not seem to enforce a filesystem
	 * name character set.
	 */
	vfstype = strdup(pvfstype);
	if (vfstype == NULL)
		perr("strdup(pvfstype)");
	for (i = 0; i < (int)strlen(vfstype); i++) {
		vfstype[i] = tolower(vfstype[i]);
		if (vfstype[i] == ' ')
			vfstype[i] = '_';
	}

	extra = getoptions(vfstype);
	optbuf = NULL;
	if (options)
		catopt(&optbuf, options);
	if (extra)
		catopt(&optbuf, extra);
	if (auxopt)
		catopt(&optbuf, auxopt);
	else if (flags & DO_BACKGRD)
		catopt(&optbuf, "-B");

	maxargc = 64;
	argv = emalloc(sizeof(char *) * maxargc);

	(void) snprintf(execbase, sizeof(execbase), "fsck_%s", vfstype);
	argc = 0;
	argv[argc++] = execbase;
	if (optbuf)
		mangle(optbuf, &argc, &argv, &maxargc);
	argv[argc++] = spec;
	argv[argc] = NULL;

	if (flags & (CHECK_DEBUG|CHECK_VERBOSE)) {
		(void)printf("start %s %swait", mntpt, 
			pidp ? "no" : "");
		for (i = 0; i < argc; i++)
			(void)printf(" %s", argv[i]);
		(void)printf("\n");
	}

	switch (pid = vfork()) {
	case -1:				/* Error. */
		warn("vfork");
		if (optbuf)
			free(optbuf);
		free(vfstype);
		return (1);

	case 0:					/* Child. */
		if ((flags & CHECK_DEBUG) && auxopt == NULL)
			_exit(0);

		/* Go find an executable. */
		execvP(execbase, _PATH_SYSPATH, __DECONST(char * const *, argv));
		if (spec)
			warn("exec %s for %s in %s", execbase, spec, _PATH_SYSPATH);
		else
			warn("exec %s in %s", execbase, _PATH_SYSPATH);
		_exit(1);
		/* NOTREACHED */

	default:				/* Parent. */
		if (optbuf)
			free(optbuf);

		free(vfstype);

		if (pidp) {
			*pidp = pid;
			return 0;
		}

		if (waitpid(pid, &status, 0) < 0) {
			warn("waitpid");
			return (1);
		}

		if (WIFEXITED(status)) {
			if (WEXITSTATUS(status) != 0)
				return (WEXITSTATUS(status));
		}
		else if (WIFSIGNALED(status)) {
			warnx("%s: %s", spec, strsignal(WTERMSIG(status)));
			return (1);
		}
		break;
	}

	return (0);
}


static int
selected(const char *type)
{
	struct entry *e;

	/* If no type specified, it's always selected. */
	TAILQ_FOREACH(e, &selhead, entries)
		if (!strncmp(e->type, type, MFSNAMELEN))
			return which == IN_LIST ? 1 : 0;

	return which == IN_LIST ? 0 : 1;
}


static const char *
getoptions(const char *type)
{
	struct entry *e;

	TAILQ_FOREACH(e, &opthead, entries)
		if (!strncmp(e->type, type, MFSNAMELEN))
			return e->options;
	return "";
}


static void
addoption(char *optstr)
{
	char *newoptions;
	struct entry *e;

	if ((newoptions = strchr(optstr, ':')) == NULL)
		errx(1, "Invalid option string");

	*newoptions++ = '\0';

	TAILQ_FOREACH(e, &opthead, entries)
		if (!strncmp(e->type, optstr, MFSNAMELEN)) {
			catopt(&e->options, newoptions);
			return;
		}
	addentry(&opthead, optstr, newoptions);
}


static void
addentry(struct fstypelist *list, const char *type, const char *opts)
{
	struct entry *e;

	e = emalloc(sizeof(struct entry));
	e->type = estrdup(type);
	e->options = estrdup(opts);
	TAILQ_INSERT_TAIL(list, e, entries);
}


static void
maketypelist(char *fslist)
{
	char *ptr;

	if ((fslist == NULL) || (fslist[0] == '\0'))
		errx(1, "empty type list");

	if (fslist[0] == 'n' && fslist[1] == 'o') {
		fslist += 2;
		which = NOT_IN_LIST;
	}
	else
		which = IN_LIST;

	while ((ptr = strsep(&fslist, ",")) != NULL)
		addentry(&selhead, ptr, "");

}


static void
catopt(char **sp, const char *o)
{
	char *s;
	size_t i, j;

	s = *sp;
	if (s) {
		i = strlen(s);
		j = i + 1 + strlen(o) + 1;
		s = erealloc(s, j);
		(void)snprintf(s + i, j, ",%s", o);
	} else
		s = estrdup(o);
	*sp = s;
}


static void
mangle(char *opts, int *argcp, const char ** volatile *argvp, int *maxargcp)
{
	char *p, *s;
	int argc, maxargc;
	const char **argv;

	argc = *argcp;
	argv = *argvp;
	maxargc = *maxargcp;

	for (s = opts; (p = strsep(&s, ",")) != NULL;) {
		/* Always leave space for one more argument and the NULL. */
		if (argc >= maxargc - 3) {
			maxargc <<= 1;
			argv = erealloc(argv, maxargc * sizeof(char *));
		}
		if (*p != '\0')  {
			if (*p == '-') {
				argv[argc++] = p;
				p = strchr(p, '=');
				if (p) {
					*p = '\0';
					argv[argc++] = p+1;
				}
			} else {
				argv[argc++] = "-o";
				argv[argc++] = p;
			}
		}
	}

	*argcp = argc;
	*argvp = argv;
	*maxargcp = maxargc;
}

static const char *
getfstype(const char *str)
{
	struct diocgattr_arg attr;
	int fd, i;

	if ((fd = open(str, O_RDONLY)) == -1)
		err(1, "cannot open `%s'", str);

	strncpy(attr.name, "PART::type", sizeof(attr.name));
	memset(&attr.value, 0, sizeof(attr.value));
	attr.len = sizeof(attr.value);
	if (ioctl(fd, DIOCGATTR, &attr) == -1) {
		(void) close(fd);
		return(NULL);
	}
	(void) close(fd);
	for (i = 0; ptype_map[i].ptype != NULL; i++)
		if (strstr(attr.value.str, ptype_map[i].ptype) != NULL)
			return (ptype_map[i].name);
	return (NULL);
}


static void
usage(void)
{
	static const char common[] =
	    "[-Cdfnpvy] [-B | -F] [-T fstype:fsoptions] [-t fstype] [-c fstab]";

	(void)fprintf(stderr, "usage: %s %s [special | node] ...\n",
	    getprogname(), common);
	exit(1);
}
